libs/sdk-common/src/lnurl/specs/auth.rs

```diff
@@ -1,39 +1,46 @@
 use std::str::FromStr;
 
-use bitcoin::hashes::{hex::ToHex, sha256, Hash, HashEngine, Hmac, HmacEngine};
-use bitcoin::secp256k1::{Message, Secp256k1};
-use bitcoin::util::bip32::{ChildNumber, ExtendedPrivKey};
-use bitcoin::KeyPair;
+use bitcoin::hashes::hex::ToHex;
+use bitcoin::util::bip32::ChildNumber;
 use reqwest::Url;
 
 use crate::prelude::*;
 
+pub trait LnurAuthSigner {
+    fn derive_bip32_pub_key(&self, derivation_path: &[ChildNumber]) -> LnUrlResult<Vec<u8>>;
+    fn sign_ecdsa(&self, msg: &[u8], derivation_path: &[ChildNumber]) -> LnUrlResult<Vec<u8>>;
+    fn hmac_sha256(
+        &self,
+        key_derivation_path: &[ChildNumber],
+        input: &[u8],
+    ) -> LnUrlResult<Vec<u8>>;
+}
+
 /// Performs the third and last step of LNURL-auth, as per
 /// <https://github.com/lnurl/luds/blob/luds/04.md>
 ///
 /// Linking key is derived as per LUD-05
 /// https://github.com/lnurl/luds/blob/luds/05.md
 ///
 /// See the [parse] docs for more detail on the full workflow.
-pub async fn perform_lnurl_auth(
-    linking_keys: KeyPair,
-    req_data: LnUrlAuthRequestData,
+pub async fn perform_lnurl_auth<S: LnurAuthSigner>(
+    req_data: &LnUrlAuthRequestData,
+    signer: &S,
 ) -> LnUrlResult<LnUrlCallbackStatus> {
-    let k1_to_sign = Message::from_slice(
-        &hex::decode(req_data.k1)
-            .map_err(|e| LnUrlError::Generic(format!("Error decoding k1: {e}")))?,
-    )?;
-    let sig = Secp256k1::new().sign_ecdsa(&k1_to_sign, &linking_keys.secret_key());
+    let url = Url::from_str(&req_data.url).map_err(|e| LnUrlError::InvalidUri(e.to_string()))?;
+    let derivation_path = get_derivation_path(signer, url)?;
+    let sig = signer.sign_ecdsa(req_data.k1.as_bytes(), &derivation_path)?;
+    let xpub = signer.derive_bip32_pub_key(&derivation_path)?;
 
     // <LNURL_hostname_and_path>?<LNURL_existing_query_parameters>&sig=<hex(sign(utf8ToBytes(k1), linkingPrivKey))>&key=<hex(linkingKey)>
```

Comment:
From [LUD-04](https://github.com/lnurl/luds/blob/luds/04.md):
```suggestion
    // <LNURL_hostname_and_path>?<LNURL_existing_query_parameters>&sig=<hex(sign(hexToBytes(k1), linkingPrivKey))>&key=<hex(linkingKey)>
```